core/drivers/stm32_gpio.c

```diff
@@ -933,21 +1209,146 @@ static TEE_Result stm32_pinctrl_dt_get(struct dt_pargs *pargs,
 }
 #endif /*CFG_DRIVERS_PINCTRL*/
 
+static void stm32_gpio_get_conf_sec(struct stm32_gpio_bank *bank)
+{
+	if (bank->sec_support) {
+		clk_enable(bank->clock);
+		bank->seccfgr = io_read32(bank->base + GPIO_SECR_OFFSET);
+		clk_disable(bank->clock);
+	}
+}
+
+static void stm32_gpio_set_conf_sec(struct stm32_gpio_bank *bank)
+{
+	if (bank->sec_support) {
+		clk_enable(bank->clock);
+		io_write32(bank->base + GPIO_SECR_OFFSET, bank->seccfgr);
+		clk_disable(bank->clock);
+	}
+}
+
+static TEE_Result stm32_gpio_sec_config_resume(void)
+{
+	struct stm32_gpio_bank *bank = NULL;
+
+	STAILQ_FOREACH(bank, &bank_list, link) {
+		if (bank->rif_cfg) {
+			if (!bank->is_tdcid)
+				continue;
+
+			bank->rif_cfg->access_mask[0] = GENMASK_32(bank->ngpios,
+								   0);
+
+			apply_rif_config(bank);
+		} else {
+			stm32_gpio_set_conf_sec(bank);
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_gpio_sec_config_suspend(void)
+{
+	struct stm32_gpio_bank *bank = NULL;
+
+	STAILQ_FOREACH(bank, &bank_list, link) {
+		if (bank->rif_cfg) {
+			if (bank->is_tdcid)
+				stm32_gpio_save_rif_config(bank);
+		} else {
+			stm32_gpio_get_conf_sec(bank);
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+stm32_gpio_sec_config_pm(enum pm_op op, unsigned int pm_hint,
+			 const struct pm_callback_handle *hdl __unused)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+
+	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT))
+		return TEE_SUCCESS;
+
+	if (op == PM_OP_RESUME)
+		ret = stm32_gpio_sec_config_resume();
+	else
+		ret = stm32_gpio_sec_config_suspend();
+
+	return ret;
+}
+DECLARE_KEEP_PAGER(stm32_gpio_sec_config_pm);
+
+/*
+ * Several pinctrl nodes can be probed. Their bank will be put in the unique
+ * bank_list. To avoid multiple configuration set for a bank when looping
+ * over each bank in the bank list, ready is set to true when a bank is
+ * configured. Therefore, during other bank probes, the configuration won't
+ * be set again.
+ */
+static TEE_Result apply_sec_cfg(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct stm32_gpio_bank *bank = NULL;
+
+	STAILQ_FOREACH(bank, &bank_list, link) {
+		if (bank->ready)
+			continue;
+
+		if (bank->rif_cfg) {
+			res = apply_rif_config(bank);
+			if (res) {
+				EMSG("Failed to set GPIO RIF, bank %u",
+				     bank->bank_id);
+				STAILQ_REMOVE(&bank_list, bank, stm32_gpio_bank,
+					      link);
+				free(bank);
```

Comment:
`bank` holds various pointers that would also need to be free. I think we could simply return with`res` here at let caller deal with it.